### ISSUE TITLE
<references> tag w/ greedy regex

### ISSUE THREAD

**Issue by xeladotbe** [spencermountain/wtf_wikipedia#148]
observed:
- incomplete document/sections (stops at list item 1.4)

expected:
- complete document/sections (including list item 11)

reproduce:

```
(async () => {
  const wtf = require('wtf_wikipedia');
  const document = await wtf.fetch('Tiger', 'de');

  document.sections().forEach(section => console.log(section.title() || 'Intro'));
})();
```

Bug caused by:
https://github.com/spencermountain/wtf_wikipedia/blob/master/src/document/preProcess/kill_xml.js#L6

I changed the regex to:

`  wiki = wiki.replace(/< ?(table|code|score|data|categorytree|charinsert|hiero|imagemap|inputbox|math|nowiki|poem|references|source|syntaxhighlight|timeline) ?[^>]{0,200}?>[\s\S]*< ?\/ ?\1 ?>/gi, ' '); // <table name=""><tr>hi</tr></table>
`

and everything seem to work as expected.

WIkipedia URL:
https://de.wikipedia.org/wiki/Tiger

Raw Wikitext URL:
https://de.wikipedia.org/w/api.php?action=query&redirects=true&prop=revisions&rvprop=content&maxlag=5&format=json&origin=*&titles=Tiger

**Comment by xeladotbe** [comment]
Update:

My modified Regexp did not work reliably. In the Wikipedia article "Zeit" there was another problem with Templates/HTML Markup.

Raw Wikitext URL:
https://de.wikipedia.org/w/api.php?action=query&redirects=true&prop=revisions&rvprop=content&maxlag=5&format=json&origin=*&titles=Zeit

I just got:
```

sentences: Array(6) [
  0: Object {
  text: "{{Infobox Physikalische Größe"
}

```

My kill_xml.js now looks like this and both articles (Zeit/Tiger) work:

```
const INVALID_TAG_NAMES = Array.of('table', 'code', 'score', 'data', 'categorytree', 
                                   'charinsert', 'hiero', 'imagemap', 'inputbox', 
                                   'math', 'nowiki', 'poem', 'references', 'source', 
                                   'syntaxhighlight', 'timeline');

const CLEANUP_EMPTY_TAGS = new RegExp('<\s*(' + INVALID_TAG_NAMES.join('|') + ')([^\/]*?)\/\s*([^>]*?)>', 'gi');
const CLEANUP_TAGS = new RegExp('<(' + INVALID_TAG_NAMES.join('|') + ')([^>]*?)>(.*?)<([^\/]*?)\/\s*\1([^>]*?)>', 'gi');

//okay, i know you're not supposed to regex html, but...
//https://en.wikipedia.org/wiki/Help:HTML_in_wikitext
const kill_xml = function(wiki) {
  //(parse <ref> tags in Section class) - luckily, refs can't be recursive.
  //other types of xml that we want to trash completely
  wiki = wiki.replace(CLEANUP_EMPTY_TAGS, ' '); // empty tags like <code />
  wiki = wiki.replace(CLEANUP_TAGS, ' '); // tags with content like <table name=''><tr><td>hi<ref>nono!</ref></td></tr></table>
  //some xml-like fragments we can also kill
  wiki = wiki.replace(/ ?< ?(span|div|table|data) [a-zA-Z0-9=" ]{2,100}\/? ?> ?/g, ' '); //<ref name="asd">
  //only kill ref tags if they are selfclosing
  wiki = wiki.replace(/ ?< ?(ref) [a-zA-Z0-9=" ]{2,100}\/ ?> ?/g, ' '); //<ref name="asd"/>
  //some formatting xml, we'll keep their insides though
  wiki = wiki.replace(/ ?<[ \/]?(p|sub|sup|span|nowiki|div|table|br|tr|td|th|pre|pre2|hr)[ \/]?> ?/g, ' '); //<sub>, </sub>
  wiki = wiki.replace(/ ?<[ \/]?(abbr|bdi|bdo|blockquote|cite|del|dfn|em|i|ins|kbd|mark|q|s)[ \/]?> ?/g, ' '); //<abbr>, </abbr>
  wiki = wiki.replace(/ ?<[ \/]?h[0-9][ \/]?> ?/g, ' '); //<h2>, </h2>
  wiki = wiki.replace(/ ?< ?br ?\/> ?/g, '\n'); //<br />
  return wiki.trim();
};
//console.log(kill_xml("hello <ref>nono!</ref> world1. hello <ref name='hullo'>nono!</ref> world2. hello <ref name='hullo'/>world3.  hello <table name=''><tr><td>hi<ref>nono!</ref></td></tr></table>world4. hello<ref name=''/> world5 <ref name=''>nono</ref>, man.}}"))
//console.log(kill_xml("hello <table name=''><tr><td>hi<ref>nono!</ref></td></tr></table>world4"))
//console.log(kill_xml('hello<ref name="theroyal"/> world <ref>nono</ref>, man}}'))
//console.log(kill_xml("hello<ref name=\"theroyal\"/> world5, <ref name=\"\">nono</ref> man"));
//console.log(kill_xml("hello <asd f> world </h2>"))
//console.log(kill_xml("North America,<ref name=\"fhwa\"> and one of"))
//console.log(kill_xml("North America,<br /> and one of"))
module.exports = kill_xml;

```

**Comment by spencermountain** [comment]
oh hey, that's another good bug, thanks for taking the time to check it out.

yeah I think it's a bad parse between the `<math>` tag used in the that section, and the nested `<references>` tag down at the bottom of the Tiger/de page. yikes.

this will reproduce it.
```js
let str = `
=== one ===
hello
<math>foo</math>
==two==
lkjsdf
==three==
<references>
</references>
`;

wtf(str).sections().forEach(section => console.log(section.title() || 'Intro'));
```

looks like a pretty-sloppy xml parse. maybe it needs a re-think. I'm glad we've found this.
lemme figure out the best solution.
cheers

**Comment by xeladotbe** [comment]
I would suggest adding a dependency for cheerio,  read the wikitext as HTML and delete all superfluous tags

**Comment by spencermountain** [comment]
ha yep, this is exactly what i was thinking, and exactly what i was hoping to avoid doing. ;)
lemme think on it - i mean, worst-case, if it's running client-side, it could be thrown into the dom too. but that all feels a little crazy for these tiny xml snippets.
i'll take a swing at it this afternoon..

**Comment by xeladotbe** [comment]
:) 

I tested a lot of articles and my latest regexp seem to work well, maybe you could just use it

**Comment by spencermountain** [comment]
hey, got this working, i went with a compromised solution. it's conceivable that any of [these](https://github.com/spencermountain/wtf_wikipedia/blob/4c51343d1d374a02fd8d62472afe5f924729071a/src/document/preProcess/kill_xml.js#L6) consecutive xml tags, within a section, will overlap and be mistreated by the regex, but i say this is fine for the timebeing.
fixed in **5.0.1**
thanks a bunch for your help
